### PR TITLE
Lint rule that would fail for PRs containing subdmodule updates #74326

### LINKED ISSUE #74326: Lint rule that would fail for PRs containing subdmodule updates
### 🚀 The feature, motivation and pitch

When submodules gets updated after checkout, devs often forget to run `git submodule update` command and propose PR that effectively reverts the update.

Prevent those accidental updates by requiring one to add "submodule-update" keyword to PR description before change can be merged.

### Alternatives

_No response_

### Additional context

_No response_

cc @seemethere @malfet @pytorch/pytorch-dev-infra

### PR DESCRIPTION
**Pull request by k-bishop**
### The feature and motivation

When submodules gets updated after checkout, devs often forget to run git submodule update command and propose PR that effectively reverts the update.

Prevent those accidental updates by requiring one to add "submodule-update" keyword to PR description before change can be merged. This allows intentional submodule updates to still make it through the linter.

Fixes #74326

### Testing

1. Create branch `behind` in personal repo (identical to this branch)
2. on `behind` branch, revert cudnn_frontend submodule to an earlier commit
3. Create PR where HEAD=`behind`, base=this branch, description="ignorance is bliss"
4. See that lint workflow fails on the new "Ensure submodules are up to date" step of quick-checks (output below):
```
Run if git diff --name-only $GH_BASE HEAD | grep -q third_party  && ! echo $PR_BODY | grep -q submodule-update; then echo "PR contains a submodule update, did you forget to run \`git submodule update\`? If this was on purpose, please specify \"submodule-update\" in the PR description."; false; fi
  if git diff --name-only $GH_BASE HEAD | grep -q third_party  && ! echo $PR_BODY | grep -q submodule-update; then echo "PR contains a submodule update, did you forget to run \`git submodule update\`? If this was on purpose, please specify \"submodule-update\" in the PR description."; false; fi
  shell: /bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.10.4/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.4/x64/lib
    PR_BODY: ignorance is bliss
    GH_BASE: 728943d6a516e7665ee1b18d162542e2b135a954
PR contains a submodule update, did you forget to run `git submodule update`? If this was on purpose, please specify "submodule-update" in the PR description.
Error: Process completed with exit code 1.
```
6. edit PR description to be "i swear I ran submodule-update"
7. Rerun workflow, see that new step now passes (output below):
```
Run if git diff --name-only $GH_BASE HEAD | grep -q third_party  && ! echo $PR_BODY | grep -q submodule-update; then echo "PR contains a submodule update, did you forget to run \`git submodule update\`? If this was on purpose, please specify \"submodule-update\" in the PR description."; false; fi
  if git diff --name-only $GH_BASE HEAD | grep -q third_party  && ! echo $PR_BODY | grep -q submodule-update; then echo "PR contains a submodule update, did you forget to run \`git submodule update\`? If this was on purpose, please specify \"submodule-update\" in the PR description."; false; fi
  shell: /bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.10.4/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.4/x64/lib
    PR_BODY: i swear i ran submodule-update
    GH_BASE: 728943d6a516e7665ee1b18d162542e2b135a954
```

### CODE REVIEW

**Review by seemethere — APPROVED** [review]
LGTM! Thanks for the contribution!

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -97,6 +97,12 @@ jobs:
         if: always()
         run: |
           (! git --no-pager grep -In '#!.*python$' -- . || (echo "The above lines have versionless Python shebangs; please specify either python2 or python3"; false))
+      - name: Ensure submodules are up to date
+        env:
+          PR_BODY: ${{ github.event.pull_request.body }}
+          GH_BASE: ${{ github.event.pull_request.base.sha }}
+        if: always()
```
Comment (seemethere) [inline .github/workflows/lint.yml:104]: Should we actually change this to only run on pull requests? I feel as though this might fail when we land to trunk since the contents of `PR_BODY` and `GH_BASE` would be blank on `push` events

**Review by seemethere — CHANGES_REQUESTED** [review]

File: .github/workflows/lint.yml (hunk 1 above)
Comment (seemethere) [reply .github/workflows/lint.yml:104]:
```suggestion
        if: ${{ github.event_name == 'pull_request' }}
```

Comment (k-bishop) [reply .github/workflows/lint.yml:104]: makes sense to me. i tested it using `on: pull_request` only, so let me make this change and confirm the test cases still pass.

**Review by janeyx99 — COMMENTED** [review]
Agreed with @seemethere's suggestion. Once you make the change this should be good to go :)
 
Also, yes the XLA failure was an upstream error that doesn't relate to your change. The new CI triggered after your change should be green now as we've temporarily skipped XLA. (There is a rerun from failed feature in GHA from the workflow page, but I don't think rerunning would have helped as it locks the base SHA which was broken.)

**Comment by janeyx99** [comment]
@k-bishop any updates on this?

**Comment by k-bishop** [comment]
> @k-bishop any updates on this?

Yes, sorry for the delay. I can make the fix tonight or tomorrow.

**Comment by k-bishop** [comment]
One thing I noticed: even with the change (i.e. run on PRs only), sometimes the base isn't picked up correctly (`fatal: bad object 9c66fc008a002d54f605c81abbc7f251aec917bd`) but this doesnt fail the workflow. I can look into why this is.

**Comment by k-bishop** [comment]
After doing some digging, the error occurs when some commit in `base` is not in `head` (i.e. when `head` is not rebased against the latest in `base`). I think I can fix this by first fetching `base` (since the gh action executes in the `head` context) and then diff-ing against `base`, but (a) fetching takes a minute or so and (b) the bash logic gets a bit hard to follow. So far I have:
```
- name: Ensure submodules are up to date
  env:
    PR_BODY: ${{ github.event.pull_request.body }}
    BASE_REF: ${{ github.event.pull_request.base.ref }}
  if: ${{ github.event_name == 'pull_request' }}
  run: |
    git fetch origin $GH_REF:$GH_REF
    (   
      changed_files=$(git diff --name-only $GH_REF HEAD)
      status=$(echo $?)
      if [ $status == 0 ]
      then 
        echo $changed_files | grep -q third_party
      else exit $status
      fi
    ) && (
      echo "$PR_BODY" | grep -q "submodule-update"
    )
    exit_code=$(echo $?)
    if [ $exit_code -eq 1 ]
    then echo "PR contains a submodule update, did you forget to run \`git submodule update\`? If this was on purpose, please specify \"submodule-update\" in the PR description.";
    fi
    exit $exit_code
```
It doesn't yet handle the case where only 1 of the predicates is true (e.g. no submodules updated, but "submodule-update" is in the PR body), but I'll keep working on it.

**Comment by janeyx99** [comment]
@k-bishop thanks for the update!

**Comment by k-bishop** [comment]
The latest commit was working in local testing (running the shell commands on my own machine) but fails in the gh actions context. Pushing here for visibility but I need to dig into why its failing

**Comment by janeyx99** [comment]
> The latest commit was working in local testing (running the shell commands on my own machine) but fails in the gh actions context. Pushing here for visibility but I need to dig into why its failing

Huh I am struggling to discern where it went wrong in CI other than just returning status 1
